compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1821,10 +1852,55 @@ impl<'a> Parser<'a> {
         let (fields, _trailing, _recovered) = self.parse_seq_to_before_end(
             &TokenKind::CloseDelim(Delimiter::Parenthesis),
             seq_sep,
-            Parser::parse_field_name,
+            |this| {
+                let res = if let token::Literal(token::Lit { kind: token::Float, symbol, suffix }) =
+                    this.token.kind
+                {
+                    match this.break_up_float(symbol) {
+                        // 1e2
+                        DestructuredFloat::Single(sym, sp) => {
+                            this.bump();
+                            thin_vec![Ident::new(sym, sp)]
+                        }
+                        // 1.
+                        DestructuredFloat::TrailingDot(sym, sym_span, dot_span) => {
+                            assert!(suffix.is_none());
+                            // Analogous to Self::break_and_eat
+                            this.token_cursor.break_last_token = true;
+                            // This might work, in cases like `1. 2.3`, and might not,
+                            // in cases like `offset_of!(Ty, 1.)`.
```

Comment:
How does `parse_expr_tuple_field_access` deal with these cases?

It seems to me that these function should be able to share more code.
E.g. `parse_expr_tuple_field_access` adds a field to a base expression, and its equivalent here could add a field to a "base vector of previous fields".
Maybe then this code and `parse_expr_tuple_field_access_float` could be merged together.

Reply:
This match arm has to split up the `1.` token into its two pieces, and then make the rest of the parser (or at least the parts that process the next few tokens) think that there is indeed these two tokens present. `parse_expr_tuple_field_access` deals with this via a `next_token` parameter, so we can just pass the dot token to it. In the case of `offset_of`, we are using [`parse_seq_to_before_end` which internally calls `Parser::expect`](https://doc.rust-lang.org/1.70.0/nightly-rustc/src/rustc_parse/parser/mod.rs.html#846), so we have to call `bump_with` instead,. which is compatible with `Parser::expect`. The precedent for that is found in [`Parser::break_and_eat`](https://doc.rust-lang.org/1.70.0/nightly-rustc/src/rustc_parse/parser/mod.rs.html#760).

Admittedly, `builtin # offset_of(Foo, 1.)` doesn't error right now, which I haven't debugged yet, but it is also not that critical imo, given that the macro layer catches this case so `offset_of!(Foo, 1.)` does error.

You can maybe put the `Token`s instead of the `Symbol` and `Span` into the `DestructuredFloat` enum, which would increase the amount of shared code slightly. But it would also make that enum sometimes pass tokens, sometimes symbols and spans, which would be weird imo. But please write if you prefer that.

Reply:
Ok, if they are hard to merge then it's fine as is.